src/update_flash.c

```diff
@@ -322,27 +366,24 @@ static int wolfBoot_delta_update(struct wolfBoot_image *boot,
         sector++;
     }
     ret = 0;
-    while((sector * WOLFBOOT_SECTOR_SIZE) < WOLFBOOT_PARTITION_SIZE) {
+
+    /* erase to the last sector */
+    while((sector * WOLFBOOT_SECTOR_SIZE) < WOLFBOOT_PARTITION_SIZE - WOLFBOOT_SECTOR_SIZE) {
         hal_flash_erase(WOLFBOOT_PARTITION_BOOT_ADDRESS +
                 sector * WOLFBOOT_SECTOR_SIZE, WOLFBOOT_SECTOR_SIZE);
         sector++;
     }
-    st = IMG_STATE_TESTING;
-    wolfBoot_set_partition_state(PART_BOOT, st);
-    /* On success, reset all flags on update partition */
-    wb_flash_erase(update, WOLFBOOT_PARTITION_SIZE - WOLFBOOT_SECTOR_SIZE,
-            WOLFBOOT_SECTOR_SIZE);
-out:
-    wb_flash_erase(swap, 0, WOLFBOOT_SECTOR_SIZE);
-#ifdef EXT_FLASH
-    ext_flash_lock();
-#endif
-    hal_flash_lock();
 
-/* Save the encryption key after swapping */
-#ifdef EXT_ENCRYPTED
-    wolfBoot_set_encrypt_key(key, nonce);
-#endif
+    /* copy the last sector of boot  which has the encrypt key and partition
```

Comment:
Doesn't this leak the key in clear to the external device (assuming EXT_SWAP=1)? Would this not defeat the purpose of the key itself?

Reply:
I wasn't aware the key couldn't be put into a potentially external swap, I thought the point of encrypted partitions was to protect the new partition in transit but I guess this protects it at rest, though I don't see why someone with physical access to the device wouldn't just read it from boot after it was decrypted.

Anyways I'm not going to argue with the referee, I've changed it to the scheme we discussed where the first sector of boot is swapped out and used to store a backup of the key when the final swap takes place. This is only necessary when encrypted partitions are used but using this final swap scheme also protects against the case where the power fails right after erasing the last boot sector. In that scenario the testing flag was never being set and the testing phase never happened.